core/src/broadcast_stage/broadcast_utils.rs

```diff
@@ -26,45 +30,65 @@ pub struct UnfinishedSlotInfo {
     pub parent: Slot,
 }
 
-/// This parameter tunes how many entries are received in one iteration of recv loop
-/// This will prevent broadcast stage from consuming more entries, that could have led
-/// to delays in shredding, and broadcasting shreds to peer validators
-const RECEIVE_ENTRY_COUNT_THRESHOLD: usize = 8;
-
 pub(super) fn recv_slot_entries(receiver: &Receiver<WorkingBankEntry>) -> Result<ReceiveResults> {
+    let target_serialized_batch_byte_count: u64 =
+        32 * ShredData::capacity(/*merkle_proof_size*/ None).unwrap() as u64;
+    let mut serialized_batch_byte_count: u64 = 8; // Vec len
     let timer = Duration::new(1, 0);
     let recv_start = Instant::now();
+
     let (mut bank, (entry, mut last_tick_height)) = receiver.recv_timeout(timer)?;
 
+    serialized_batch_byte_count += serialized_size(&entry)?;
     let mut entries = vec![entry];
-    let mut slot = bank.slot();
-    let mut max_tick_height = bank.max_tick_height();
-
-    assert!(last_tick_height <= max_tick_height);
-
-    if last_tick_height != max_tick_height {
-        while let Ok((try_bank, (entry, tick_height))) = receiver.try_recv() {
-            // If the bank changed, that implies the previous slot was interrupted and we do not have to
-            // broadcast its entries.
-            if try_bank.slot() != slot {
-                warn!("Broadcast for slot: {} interrupted", bank.slot());
-                entries.clear();
-                bank = try_bank;
-                slot = bank.slot();
-                max_tick_height = bank.max_tick_height();
-            }
-            last_tick_height = tick_height;
-            entries.push(entry);
-
-            if entries.len() >= RECEIVE_ENTRY_COUNT_THRESHOLD {
-                break;
-            }
-
-            assert!(last_tick_height <= max_tick_height);
-            if last_tick_height == max_tick_height {
-                break;
-            }
+
+    assert!(last_tick_height <= bank.max_tick_height());
+
+    // Drain channel
+    while last_tick_height != bank.max_tick_height()
+        && serialized_batch_byte_count < target_serialized_batch_byte_count
```

Comment:
```
serialized_batch_byte_count < target_serialized_batch_byte_count
```
causes this loop to break  before channel is drained.
This first loop does not need to track or inspect size of entries.